client/core/core.go

```diff
@@ -413,29 +423,58 @@ func (dc *dexConnection) processNomatch(oid order.OrderID) error {
 }
 
 // runMatches runs the sorted matches returned from parseMatches.
-func (dc *dexConnection) runMatches(matches map[order.OrderID]*serverMatches) error {
-	for _, match := range matches {
-		tracker := match.tracker
-		if len(match.msgMatches) > 0 {
-			err := tracker.negotiate(match.msgMatches)
+func (dc *dexConnection) runMatches(tradeMatches map[order.OrderID]*serverMatches) error {
+	runMatch := func(sm *serverMatches) error {
+		tracker := sm.tracker
+		errs := newErrorSet("trade %v - ", sm.tracker.ID())
+		// Begin negotiation for any msgMatches.
+		if len(sm.msgMatches) > 0 {
+			err := tracker.negotiate(sm.msgMatches)
 			if err != nil {
-				return err
-			}
-			_, err = tracker.tick()
-			if err != nil {
-				return err
+				errs.add("negotiate: %v", err)
+				// Go on to process any match.cancel.
+			} else {
+				// Try to tick the trade now, but do not interrupt on error. The
+				// trade will tick again automatically.
+				_, err = tracker.tick()
+				if err != nil {
+					errs.add("tick: %v", err)
+				}
 			}
-
 		}
-		if match.cancel != nil {
-			err := tracker.processCancelMatch(match.cancel)
+
+		// Record any cancel of this trade.
+		if sm.cancel != nil {
+			err := tracker.processCancelMatch(sm.cancel)
 			if err != nil {
-				return err
+				errs.add("processCancelMatch for cancel order %v: %v",
+					sm.cancel.OrderID, err)
 			}
 		}
+
+		return errs.ifAny()
 	}
+
+	// Process the trades concurrently.
+	errChan := make(chan error)
+	for _, trade := range tradeMatches {
+		go func(trade *serverMatches) {
+			errChan <- runMatch(trade)
+		}(trade)
+	}
```

Comment:
`negotiate` still handles a trade's matches the same way, but the trades in `runMatches` are processed concurrently.